### PR TITLE
[Arm][AArch64][Clang] Respect function's branch protection attributes.

### PR DESCRIPTION
**Pull request by DanielKristofKiss**
Default attributes assigned to all functions according to the command line parameters. Some functions might have their own attributes and we need to set or remove attributes accordingly.
Tests are updated to test this scenarios too.

### CODE REVIEW

File: clang/lib/CodeGen/TargetInfo.cpp [hunk 1]
```diff
@@ -209,9 +209,28 @@ llvm::Value *TargetCodeGenInfo::createEnqueuedBlockKernel(
...
+  };
+
+  AddRemoveAttributeAsSet(BPI.BranchTargetEnforcement,
+                          "branch-target-enforcement");
+  AddRemoveAttributeAsSet(BPI.BranchProtectionPAuthLR,
+                          "branch-protection-pauth-lr");
+  AddRemoveAttributeAsSet(BPI.GuardedControlStack, "guarded-control-stack");
 }
 
 void TargetCodeGenInfo::setBranchProtectionFnAttributes(
```
Comment (stuij) [inline clang/lib/CodeGen/TargetInfo.cpp:236]: I can see that this AttrBuilder version of the fn is still used in `getTrivialDefaultFunctionAttributes` in `CGCall.cpp`. I guess in practice the outcome would still be the same for this use case between old and new implementation, but even if I didn't miss something it does feel a bit icky to do so.

Perhaps add a comment to explain the code duplication?

Comment (DanielKristofKiss) [reply clang/lib/CodeGen/TargetInfo.cpp]: Thanks for the concern. I added more comments and renamed the AttrBuilder version to make it more distinct.

**Review by jthackray — APPROVED** [review]
LGTM.

**Review by stuij — COMMENTED** [review]
LGTM, thanks!

**Comment by RSilicon** [comment]
Who here can add milestones?

**Comment by RSilicon** [comment]
> 9e9fa00

It has a merge conflict too, but one that is trivial so I can make my own PR.

**Comment by RSilicon** [comment]
https://github.com/llvm/llvm-project/pull/102646

**Pull request merged**